text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+  for backwards compatibility) will also be reimplemented. For example, to a class like `CdkAppWorkflow`.
+
+We will gain efficiency from the fact that the reimplementation will not contain business logic: it will
+mostly be mapping types and properties between the different levels. All the knowledge about the structure
+and order of deploying CDK apps will be located in layer 2, and is reused.
+
+### Why will we stay backwards compatible?
+
+Many people have already invested a lot of effort into writing code against the CDK Pipelines
+library. The changes we are planning mostly affect the internals of the library, and we want to add
+some slight enhanced expressiveness and customizability to the API.
+
+With a little additional effort, we can make the new API additive and have the old one continue to work,
+giving people the opportunity to switch over to the new API at their own pace.
+
+### How will we stay backwards compatible?
+
+Right now, the parts that are too limiting are the parts that are directly re-exposed from the `aws-codepipeline`
+library. For example, the fact that the synth needs to be a `codepipeline.IAction`.
+
+What we'll do is type testing to switch between old and new behavior. Conceptually:
+
+```ts
+interface IModernSynthAction implements codepipeline.IAction {
+  // Or whatever the signature ends up being
+  public addToPipelineModern(pipeline: CdkPipeline): AddResult;
+}
+
+if (isModernSynthAction(props.synthAction)) {
+  // New behavior
+  props.synthAction.addToPipelineModern(this);
+} else {
+  // Wrapper class for legacy behavior
+  new LegacySynthAction(props.synthAction).addToPipelineModern(this);
+}
+```
+
+### How do we render deployments out across backends?
+
+In the middle layer, a stack deployments gets represented as nested state machines,
+like this:
+
+```
+┌───────────────────────────────────────────────────────────────┐
+│                          DeployStack                          │
+│                                                               │
+│     ┌────────────────────┐         ┌────────────────────┐     │
+│     │                    │         │                    │     │
+│     │  CreateChangeSet   │────────▶│  ExecuteChangeSet  │     │
+│     │                    │         │                    │     │
+│     └────────────────────┘         └────────────────────┘     │
+│                                                               │
+└───────────────────────────────────────────────────────────────┘
+```
+
+A backend rendered has the choice to either:
+
+* Recognize a `DeployStack` state and render that out to a `cdk deploy` command (or something
+  else appropriate); or
+* Recognize the individual `CreateChangeSet` and `ExecuteChangeSet` states and render those
+  out either to CodePipeline actions or `aws cloudformation change-change-set` CLI commands,
+  as may be appropriate.
```

Comment:
Feels a little fragile. The backend should either respect the full model or not. Otherwise we may end up with features that work only in specific backends...

I am wondering if "DeployStack" is a sufficient level of detail for the middle layer. What value do we get by modeling "CreateChangeSet"  and "ExecuteChangeSet" in the middle tier? Feels like this is an implementation detail that each backend should determine.